### DISCUSSION TITLE
Cannot get TemplatedConfigLoader + jinja to work

### DISCUSSION THREAD

**Discussion by mjmare** [kedro-org/kedro#1243] [Idea]
I'm a Kedro newb and need some help getting TemplatedConfigLoader with a catalog.yml that contains a ninja snippet to work. The relevant parts (mostly lifted from the docs):

hooks.py:
```
class ProjectHooks:
    @hook_impl
    def register_config_loader(self, conf_paths: Iterable[str]) -> ConfigLoader:
        return TemplatedConfigLoader(
            conf_paths,
            globals_pattern="*globals.yml",  # read the globals dictionary from project config
            globals_dict={},  # extra keys to add to the globals dictionary, take precedence over globals_pattern
        )
```

conf/base/globals.yml
```
my_tables:
  - AA
  - BB
  - CC
```

part of catalog.yml
```
:
:

{% for table in my_tables %}
{{ table }}:
  layer: primary
  type: pandas.ParquetDataSet
  filepath: data/03_primary/{{table}}.parquet
{% endfor %}
```

I don't see any errors (but my pipeline does not acces the new tables yet). In some debugging sessions I see globals.yml being accessed. 
However when using Kedro ipython to use the catalog.list(), the newly created data tables are not there. As if the loop is skipped. 

Sure enough, when I change the first jinja line to:
```
{% for table in my_tables|default(['ZZ', 'XX']) %}
```
I see two tables 'ZZ', and 'XX' appear in the catalog. So it seems my_tables is empty.

Conclusion: maybe my understanding of globals_pattern and globals_dict is wrong.  I thought they were used to populate a dict that was passed to the jinja template to use as global variables. 

Please elucidate me. TIA

**Comment by datajoely** [comment]
Unfortunately you can't mix Kedro Templated globals and Jinja scoped variables out of the box. An open source user on our discord documented how to do this here:
https://discord.com/channels/778216384475693066/846330075535769601/933658953773101078

**Reply by mjmare** [reply]
So it is by design, huh? Didn't expect that. 
It appears to be on the roadmap though (as part of a bigger config overhaul). 

Thanks for the pointer.

**Reply by datajoely** [reply]
@mjmare - yes jinja especially is something we enabled as a temporary solution. It's something we we're going to improve in 0.18.x .

**Reply by mjmare** [reply]
Temporary? It can go away then? 
I find the ability to script the catalog (or any config) to be very useful.

**Reply by datajoely** [reply]
Sorry I should make myself clearer - there is a big piece of research going on on issue #891 

We may end up keeping Jinja support. Fundamentally iti is a solution to Kedro requiring too much config at a certain scale. This then begs the question, is this the right problem being solved? Can we find a way to reduce the config overhead on users? The Google SRE team did a really great piece of [research on this topic](https://sre.google/workbook/configuration-specifics/) and they highlight similar concerns.

In any case - we will likely find a way of introducing our new approach without breaking the old one, but hopefully this gives you a view of our current thinking.

**Reply by mjmare** [reply]
I'm glad to see that the Kedro team is spending considerable effort to make our config life better. ;-)
Curious what the outcome will be.

**Reply by gianniwgm** [reply]
I can't see the content of the discord link and I have the exact same issue. Could you share the solution in another way? Thanks

**Reply by mjmare** [reply]
I had the same problem initially with Discord (totally confusing UI): you have to press some small green checkmark to accept the terms. 
Anyway here is the code for future reference: 
```
class TemplatedConfigLoaderGlobals(TemplatedConfigLoader):
    def _load_config_file(self, config_file: Path) -> Dict[str, Any]:
        """Load an individual config file using `anyconfig` as a backend.
        Args:
            config_file: Path to a config file to process.
        Returns:
            Parsed configuration.
        """
        # for performance reasons
        import anyconfig  # pylint: disable=import-outside-toplevel

        if "globals" in str(config_file):
            globals = {}
        else:
            globals_pattern = "*globals.yml"
            globals = self.get(globals_pattern) if globals_pattern else {}

        return {
            k: v
            for k, v in anyconfig.load(
                config_file, ac_template=True, ac_context=globals
            ).items()
            if not k.startswith("_")
        }

class ProjectHooks:
    @hook_impl
    def register_config_loader(self, conf_paths: Iterable[str]) -> ConfigLoader:
        return TemplatedConfigLoaderGlobals(
            conf_paths,
            globals_pattern="*globals.yml",  # read the globals dictionary from project config
            globals_dict={},  # extra keys to add to the globals dictionary, take precedence over globals_pattern
        )
```

**Reply by datajoely** [reply]
Thanks @mjmare ! the trick to this is the `ac_context=globals` part which isn't in our current class

**Comment by mjmare** [comment]
How does this work in v0.18.0? My understanding is that the register_config_loader hook has gone away. Does TemplatedConfigLoader support this out of the box?

**Reply by datajoely** [reply]
So you need to define it in `settings.py` but there is a bug with a workaround fix here https://github.com/kedro-org/kedro/issues/1402

**Reply by mjmare** [reply]
No, I have TemplatedConfigLoader working with the said hack, but the templated config load still does not load the variables in globals.yml into the **jinja context**.
Since the implementation of the TemplatedConfigLoader has changed to old solution does not work anymore.

**Reply by datajoely** [reply]
This should be fixed shortly in 0.18.1

**Reply by datajoely** [reply]
#1437

**Reply by mjmare** [reply]
Thanks. What's the ETA for 0.18.1?

**Reply by merelcht** [reply]
We're aiming to release at the end of this sprint, so in approximately 2 weeks.

**Reply by mjmare** [reply]
OK, thx
